perf: Use SmallVec in the Type::App variant

Marwes opened the pull request:
Since applications are usually done on just a few argument it is usually unnecessary to allocate a `Vec`. Using `SmallVec` this heap allocation can be avoided for applications with only a few arguments which should cover most of the cases.
#50

Marwes commented:
Not as happy as I expected with this PR as `SmallVec` currently has two drop flags which causes `Type` to grow by 8 bytes. Once https://github.com/rust-lang/rust/pull/35764 gets merged it should be possible to actually increase the number of arguments to 3 (from the current 2) while still keeping the same size for `Type` (thought naturally this would be worse for anything else than nightly until the trains catch up).

Atm I am only seeing a minor improvement  in the `typecheck_prelude` benchmark so maybe hold of on merging this until it can be tested with the aforementioned rustc PR.

lfairy commented:
Heads up, rust-lang/rust#35764 was merged a while ago. Might like to revisit this change.

Marwes commented:
I think I redid the benchmarks after rust-lang/rust#35764 was merged but didn't see any significant improvements (allocation does not seem to be a large part of the runtime currently, but admittedly it might be shadowed by other things). Going to redo it anyway to verify (and remember to update this PR).

I believe drop-flags will only be removed when 1.13 is released however so In any case it might be best waiting for 1.13 before merging.

Marwes commented:
Rebased and removed the intermediate `Vecs` used when constructing `AppVecs` (which gave an additional ~5% speedup). Used the master branch of https://github.com/servo/rust-smallvec so we should wait with merging this until at least a new version on cargo is released. May also want to wait for rust 1.13 as this will likely be a regression in performance on stable otherwise.

# SmallVec
```
test typecheck_prelude ... bench:  13,559,631 ns/iter (+/- 569,445)
```

# Vec
```
test typecheck_prelude ... bench:  15,035,660 ns/iter (+/- 570,014)
```

brendanzab commented on check/src/unify_type.rs:
```diff
@@ -180,16 +182,22 @@ fn do_zip_match<'a, U>(self_: &ArcType,
-                        unifier.try_match(l, &Type::app(r.clone(), r_args[..offset].into()));
+
+                    let reduced_r = Type::app(r.clone(),
+                                              r_args[..offset].iter().cloned().collect());
```
Could you make this an intermediate? like:

```rust
let offset = r_args.len() - l_args.len();

let reduced_r_args = r_args[..offset].iter().cloned();
let reduced_r = Type::app(r.clone(), reduced_r_args.collect());
```

Just find the weird indentation kinda ugly.

Marwes replied:
I find it hard to read with to many superfluous intermediates as well (neither of the names `reduced_r` or `reduced_r_args` says anything). I will see what I can do, may be able to simply the code overall.

brendanzab replied:
Agreed...  no real 'best way' here. Not a big deal - I agree it might be a tad superfluous here.

_shakes fist at Rust's ugly syntax and rightward drift_

😆

brendanzab commented on check/tests/pass.rs:
```diff
@@ -346,10 +348,10 @@ test
-    let variants = Type::variants(vec![(support::intern_unscoped("T"),
+    let variants = Type::variants(collect![(support::intern_unscoped("T"),
                                         Type::function(vec![typ("a")],
                                                        support::typ_a("Test", vec![typ("a")])))]);
```
Indentation is now out of alignment.

Marwes replied:
Ihave got rustfmt set to format automatically but I think `vec!` is special cased, other macros do not get formatted...

brendanzab replied:
Yep, that's right :(

brendanzab commented on base/src/types.rs:
```diff
@@ -1212,7 +1225,7 @@ fn walk_move_types2<'a, I, F, T>(mut types: I, replaced: bool, output: &mut Vec<
                 output.push(typ);
             }
-            None if replaced || !output.is_empty() => {
+            None if replaced || output.len() != 0 => {
```
Reasoning?

brendanzab replied:
Dang - yeah, a PR would probably be useful there.

brendanzab commented on base/src/types.rs:
```diff
@@ -1186,24 +1196,27 @@ pub fn walk_move_type_opt<F: ?Sized, I, T>(typ: &Type<I, T>, f: &mut F) -> Optio
+    let mut out = R::default();
     walk_move_types2(types.into_iter(), false, &mut out, &mut f);
-    if out.is_empty() {
+    if out.len() == 0 {
```
Ditto

Marwes commented:
Rebase. Since rust 1.13 is release there is no reason to delay merging this anymore I think. Disabled osx builds since they take forever to start.

brendanzab commented on base/src/types.rs:
```diff
@@ -1077,12 +1085,14 @@ pub fn walk_move_type_opt<F: ?Sized, I, T>(typ: &Type<I, T>, f: &mut F) -> Optio
 
-pub fn walk_move_types<'a, I, F, T>(types: I, mut f: F) -> Option<Vec<T>>
+// FIXME Add R: Default and remove the `out` parameter
+pub fn walk_move_types<'a, I, F, T, R>(types: I, mut f: F) -> Option<R>
```
Could you elaborate on this?

Marwes replied:
Just forgot to remove that comment when removing the out parameter, fixed.

brendanzab replied:
:D

brendanzab commented on check/src/unify_type.rs:
```diff
@@ -277,6 +255,53 @@ fn do_zip_match<'a, U>(self_: &ArcType,
+            Ok(merge(r, new_type, r_args, new_args, Type::app))
+        }
+    }
+}
```
Love this - great idea breaking it out into another function ❤️
